cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/KafkaLogProcessorPipeline.java

```diff
@@ -0,0 +1,623 @@
+        if (eventsAppended <= 0) {
+          break;
+        }
+        unFlushedEvents += eventsAppended;
+      }
+
+      try {
+        ILoggingEvent loggingEvent = serializer.fromBytes(message.message().payload());
+        if (needAppend(loggingEvent)) {
+          // Use the message payload size as the size estimate of the logging event
+          // Although it's not the same as the in memory object size, it should be just a constant factor, hence
+          // it is proportional to the actual object size.
+          eventQueue.add(loggingEvent, loggingEvent.getTimeStamp(),
+                         message.message().payloadSize(), partition, message.offset());
+        }
+      } catch (IOException e) {
+        // This should happen. In case it happens (e.g. someone published some garbage), just skip the message.
+        LOG.warn("Fail to decode logging event from {}:{} at offset {}. Skipping it.",
+                 topic, partition, message.offset(), e);
+      }
+
+      offset = message.nextOffset();
+    }
+
+    return offset;
+  }
+
+  /**
+   * Fetches messages from Kafka across all partitions simultaneously.
+   */
+  private <T extends Map<Integer, Future<Iterable<MessageAndOffset>>>> T fetchAll(Map<Integer, Long> offsets,
+                                                                                  T fetchFutures) {
+    for (final int partition : config.getPartitions()) {
+      final long offset = offsets.get(partition);
+
+      fetchFutures.put(partition, fetchExecutor.submit(new Callable<Iterable<MessageAndOffset>>() {
+        @Override
+        public Iterable<MessageAndOffset> call() throws Exception {
+          return fetchMessages(partition, offset);
+        }
+      }));
+    }
+
+    return fetchFutures;
+  }
+
+  /**
+   * Appends buffered events to appender. If the {@code force} parameter is {@code false}, buffered events
+   * that are older than the buffer milliseconds will be appended and removed from the buffer.
+   * If {@code force} is {@code true}, then at least {@code maxQueueSize * MIN_FREE_FACTOR} events will be appended
+   * and removed, regardless of the event time.
+   *
+   * @return number of events appended to the appender
+   */
+  private int appendEvents(long currentTimeMillis, boolean forced) {
+    long minEventTime = currentTimeMillis - config.getBufferMillis();
+    long maxRetainSize = forced ? (long) (config.getMaxQueueSize() * MIN_FREE_FACTOR) : Long.MAX_VALUE;
+
+    TimeEventQueue.EventIterator<ILoggingEvent, Long> iterator = eventQueue.iterator();
+
+    int eventsAppended = 0;
+    while (iterator.hasNext()) {
+      ILoggingEvent event = iterator.next();
+
+      // If not forced to reduce the event queue size and the current event timestamp is still within the
+      // buffering time, no need to iterate anymore
+      if (eventQueue.getEventSize() <= maxRetainSize && event.getTimeStamp() >= minEventTime) {
```

Comment:
The side effect of dequeuing without reaching the buffering time is out of order log events. That's why instead of dequeuing everything events, we dequeue half to maintain a balance between event ordering and memory limitation.